src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs

```diff
@@ -1242,92 +1277,127 @@ internal static bool DeepEquals(JsonElement left, JsonElement right)
                     return true;
 
                 case JsonValueKind.Number:
-                    return JsonHelpers.AreEqualJsonNumbers(left.GetRawValue().Span, right.GetRawValue().Span);
+                    return JsonHelpers.AreEqualJsonNumbers(element1.GetRawValue().Span, element2.GetRawValue().Span);
 
                 case JsonValueKind.String:
-                    if (right.ValueIsEscaped)
+                    if (element2.ValueIsEscaped)
                     {
-                        if (left.ValueIsEscaped)
+                        if (element1.ValueIsEscaped)
                         {
                             // Both values are escaped, force an allocation to unescape the RHS.
-                            return left.ValueEquals(right.GetString());
+                            return element1.ValueEquals(element2.GetString());
                         }
 
                         // Swap values so that unescaping is handled by the LHS.
-                        (left, right) = (right, left);
+                        (element1, element2) = (element2, element1);
                     }
 
-                    return left.ValueEquals(right.ValueSpan);
+                    return element1.ValueEquals(element2.ValueSpan);
 
                 case JsonValueKind.Array:
-                    ArrayEnumerator rightArrayEnumerator = right.EnumerateArray();
-                    foreach (JsonElement leftElement in left.EnumerateArray())
+                    if (element1.GetArrayLength() != element2.GetArrayLength())
+                    {
+                        return false;
+                    }
+
+                    ArrayEnumerator arrayEnumerator2 = element2.EnumerateArray();
+                    foreach (JsonElement e1 in element1.EnumerateArray())
                     {
-                        if (!rightArrayEnumerator.MoveNext() || !DeepEquals(leftElement, rightArrayEnumerator.Current))
+                        bool success = arrayEnumerator2.MoveNext();
+                        Debug.Assert(success, "enumerators must have matching length");
+
+                        if (!DeepEquals(e1, arrayEnumerator2.Current))
                         {
                             return false;
                         }
                     }
 
```

Comment:
You have an assert above that MoveNext returns true. Should we have one here that it returns false?